File: src/cpp/y_dbms.cpp
#include "ordinal.h"
#include "wy.h"
#include <algorithm>
#include <sstream>
#include <string>
#include <vector>

// ============================================================
// 1-Y / ω-Y sequence ↔ DBMS (Dimensional BMS) conversion
//
// Below [1,3]: DBMS = 0-Y → BMS with trailing zeros stripped.
// [1,3]+:      DBMS derived from ω-Y mountain with ω-boundary markers.
// ============================================================

static std::vector<int> parseSeqStr(const std::string &s) {
  std::vector<int> seq;
  std::stringstream ss(s);
  std::string item;
  while (std::getline(ss, item, ','))
    seq.push_back(std::stoi(item));
  return seq;
}

/// Check if row label transition is an ω-boundary
static bool isOmegaBoundary(const std::vector<int> &a, const std::vector<int> &b) {
  if (a.empty() || b.empty())
    return false;
  if (b.size() >= 2 && b[0] == 0) {
    int extA = (a.size() >= 2) ? a[1] : 0;
    int extB = b[1];
    return extB > extA;
  }
  if (b.size() > a.size() + 1)
    return true;
  if (b.size() > a.size() && b.back() > 0)
    return true;
  return false;
}

/// Count ω-boundaries in the ω-Y mountain row labels
static int countOmegaBoundaries(const std::vector<std::vector<int>> &wyRows) {
  int count = 0;
  for (size_t i = 1; i < wyRows.size(); i++)
    if (isOmegaBoundary(wyRows[i - 1], wyRows[i]))
      count++;
  return count;
}

/// Convert a 1-Y sequence to DBMS matrix.
/// Below [1,3]: uses zeroYToBMS.  [1,3]+: returns fixed placeholder ≥(0)(1)(2,1,,1).
Matrix oneYToDBMS(const std::vector<int> &seq) {
  if (seq.empty() || seq[0] == 0)
    return {};
  for (int v : seq)
    if (v < 0)
      return {};

  // Check for ω-boundaries using ω-Y mountain
  auto [mountain, wyRows] = buildWYMountainWithRows(seq, -1, false);
  if (mountain.empty())
    return {};
  if (countOmegaBoundaries(wyRows) > 0) {
    // No stable algorithm for [1,3]+ yet; return fixed placeholder
    Matrix placeholder;
    placeholder.push_back({0});           // (0)
    placeholder.push_back({1});           // (1)
    placeholder.push_back({2, 1, -2, 1}); // (2,1,,1)
    return placeholder;
  }

  // Below [1,3]: use 0-Y → BMS conversion
  Matrix bms = zeroYToBMS(seq);
  if (bms.empty())
    return {};
  for (auto &col : bms)
    while (col.size() > 1 && col.back() == 0)
      col.pop_back();
  return bms;
}

/// Convert a DBMS matrix to a 1-Y sequence.
std::vector<int> dbmsToOneY(const Matrix &dbms) {
  if (dbms.empty() || dbms[0].empty())
    return {};

  bool hasMarker = false;
  for (auto &col : dbms)
    for (int v : col)
      if (v == -2) {
        hasMarker = true;
        break;
      }

  if (hasMarker)
    return {};

  std::string seqStr = bmsTo0YSequence(dbms);
  if (seqStr.empty())
    return {};
  return parseSeqStr(seqStr);
}

/// Format a DBMS matrix as a readable string like (0)(1)(2,1)
std::string dbmsToString(const Matrix &dbms) {
  std::string out;
  for (size_t j = 0; j < dbms.size(); j++) {
    out += "(";
    bool first = true;
    for (size_t r = 0; r < dbms[j].size(); r++) {
      if (dbms[j][r] == -2) {
        out += ",,";
        first = true;
        continue;
      }
      if (!first)
        out += ",";
      out += std::to_string(dbms[j][r]);
      first = false;
    }
    out += ")";
  }
  return out;
}

/// Convert DBMS to standard BMS. DBMS (without ω-boundary markers)
/// is triangular BMS with variable-length columns. The conversion
/// pads columns to uniform length and then calls triangularToBMS.
Matrix dbmsToBMS(const Matrix &dbms) {
  for (auto &col : dbms)
    for (int v : col)
      if (v == -2)
        return {};
  size_t maxLen = 0;
  for (auto &col : dbms)
    if (col.size() > maxLen)
      maxLen = col.size();
  if (maxLen == 0)
    return {};
  size_t target = std::max(maxLen, size_t(2));
  Matrix padded;
  for (auto &col : dbms) {
    MatrixRow r = col;
    r.resize(target, 0);
    padded.push_back(std::move(r));
  }
  return triangularToBMS(padded);
}

File: src/rust/core/src/y_dbms.rs
//! 1-Y / ω-Y sequence ↔ DBMS (Dimensional BMS) conversion.

use crate::triangular::triangular_to_bms;
use crate::wy::build_wy_mountain_with_rows;
use crate::zero_y::{bms_to_0y_sequence, zero_y_to_bms};
use crate::Matrix;

fn parse_seq_str(s: &str) -> Vec<i32> {
    s.split(',').map(|item| item.trim().parse().unwrap_or(0)).collect()
}

/// Check if row label transition is an ω-boundary
fn is_omega_boundary(a: &[i32], b: &[i32]) -> bool {
    if a.is_empty() || b.is_empty() {
        return false;
    }
    if b.len() >= 2 && b[0] == 0 {
        let ext_a = if a.len() >= 2 { a[1] } else { 0 };
        let ext_b = b[1];
        return ext_b > ext_a;
    }
    if b.len() > a.len() + 1 {
        return true;
    }
    if b.len() > a.len() && *b.last().unwrap() > 0 {
        return true;
    }
    false
}

/// Count ω-boundaries in the ω-Y mountain row labels
fn count_omega_boundaries(wy_rows: &[Vec<i32>]) -> i32 {
    let mut count = 0;
    for i in 1..wy_rows.len() {
        if is_omega_boundary(&wy_rows[i - 1], &wy_rows[i]) {
            count += 1;
        }
    }
    count
}

/// Convert a 1-Y sequence to a DBMS matrix.
/// Below [1,3]: uses zeroYToBMS. [1,3]+: returns fixed placeholder ≥(0)(1)(2,1,,1).
pub fn one_y_to_dbms(seq: &[i32]) -> Matrix {
    if seq.is_empty() || seq[0] == 0 {
        return Vec::new();
    }
    for &v in seq {
        if v < 0 {
            return Vec::new();
        }
    }

    // Check for ω-boundaries using ω-Y mountain
    let (mountain, wy_rows) = build_wy_mountain_with_rows(seq, -1, false);
    if mountain.is_empty() {
        return Vec::new();
    }
    if count_omega_boundaries(&wy_rows) > 0 {
        // No stable algorithm for [1,3]+ yet; return fixed placeholder
        let mut placeholder: Matrix = Vec::new();
        placeholder.push(vec![0]); // (0)
        placeholder.push(vec![1]); // (1)
        placeholder.push(vec![2, 1, -2, 1]); // (2,1,,1)
        return placeholder;
    }

    // Below [1,3]: use 0-Y → BMS conversion
    let mut bms = zero_y_to_bms(seq);
    if bms.is_empty() {
        return Vec::new();
    }
    for col in &mut bms {
        while col.len() > 1 && *col.last().unwrap() == 0 {
            col.pop();
        }
    }
    bms
}

/// Convert a DBMS matrix to a 1-Y sequence.
pub fn dbms_to_one_y(dbms: &Matrix) -> Vec<i32> {
    if dbms.is_empty() || dbms[0].is_empty() {
        return Vec::new();
    }

    let mut has_marker = false;
    for col in dbms {
        for &v in col {
            if v == -2 {
                has_marker = true;
                break;
            }
        }
    }

    if has_marker {
        return Vec::new();
    }

    let seq_str = bms_to_0y_sequence(dbms);
    if seq_str.is_empty() {
        return Vec::new();
    }
    parse_seq_str(&seq_str)
}

/// Format a DBMS matrix as a readable string like (0)(1)(2,1)
pub fn dbms_to_string(dbms: &Matrix) -> String {
    let mut out = String::new();
    for col in dbms {
        out += "(";
        let mut first = true;
        for (_r, v) in col.iter().enumerate() {
            if *v == -2 {
                out += ",,";
                first = true;
                continue;
            }
            if !first {
                out += ",";
            }
            out += &v.to_string();
            first = false;
        }
        out += ")";
    }
    out
}

/// Convert DBMS to standard BMS.
pub fn dbms_to_bms(dbms: &Matrix) -> Matrix {
    for col in dbms {
        for &v in col {
            if v == -2 {
                return Vec::new();
            }
        }
    }
    let mut max_len = 0;
    for col in dbms {
        if col.len() > max_len {
            max_len = col.len();
        }
    }
    if max_len == 0 {
        return Vec::new();
    }
    let target = max_len.max(2);
    let mut padded: Matrix = Vec::with_capacity(dbms.len());
    for col in dbms {
        let mut r = col.clone();
        r.resize(target, 0);
        padded.push(r);
    }
    triangular_to_bms(&padded)
}
